python/taichi/misc/image.py

```diff
@@ -15,16 +15,15 @@ def imwrite(img, filename):
 
     assert len(img.shape) in [2,
                               3], "Image must be either RGB/RGBA or greyscale"
-    assert img.shape[2] in [1, 3,
-                            4], "Image must be either RGB/RGBA or greyscale"
 
     resx, resy = img.shape[:2]
     if len(img.shape) == 2:
         comp = 1
     else:
         comp = img.shape[2]
+    assert comp in [1, 3, 4], "Image must be either RGB/RGBA or greyscale"
 
-    img = np.ascontiguousarray(img.swapaxes(0, 1)[::-1, :, :])
+    img = np.ascontiguousarray(img.swapaxes(0, 1)[::-1, :])
```

Comment:
Honestly I felt a little hard to understand this LOC.
Would you be happy to give me some insights why we need to
1. exchange the image's axes between 0 and 1
2. then reverse the first literally "height" component. Is it related to stb_image?

Reply:
1. Because `np.ascontiguousarray` is row-major and stb_image is height-major, therefore we obtain `(height, width)`.
2. Because Taichi GUI's coordinate system is up = `1`, down = `0` for intuition, while the physical storage in memory, is up = `0`, down = `height-1`.

Reply:
Thanks for you clear explanation! Now I got it. and your pytest is so professional!

Meanwhile, It seems currently our test only cover the numpy input instead our another ti.Matrix ti.var and ti.Vector input.

Do you have any interest in tesing them functionalities for Taichi tensors?
```
@ti.host_arch_only
def test_image_io_matrix(resx, resy, comp, ext):
pixel = ti.Matrix(comp, 1, dt=ti.f32, shape = (resx, resy))

@ti.kernel
def draw():
    for i, j in pixel:
        for c in ti.static(range(comp)):
            pixel[i, j][c] = ti.random()

draw()

from tempfile import mkstemp

fn = mkstemp(suffix='.' + ext)[1]
ti.imwrite(pixel, fn)

pixel_r = ti.imread(fn).reshape(resx, resy, comp)
pixel = (pixel.to_numpy() * 255 + 0.5).astype(np.uint8)
assert (pixel_r == pixel).all()
@ti.host_arch_only
def test_image_io_vector(resx, resy, comp, ext):
pixel = ti.Vector(n=comp, dt=ti.f32, shape = (resx, resy))

@ti.kernel
def draw():
    for i, j in pixel:
        for c in ti.static(range(comp)):
            pixel[i, j][c] = ti.random()

draw()

from tempfile import mkstemp

fn = mkstemp(suffix='.' + ext)[1]
ti.imwrite(pixel, fn)

pixel_r = ti.imread(fn).reshape(resx, resy, comp)
pixel = (pixel.to_numpy() * 255 + 0.5).astype(np.uint8)
assert (pixel_r == pixel).all()
@ti.host_arch_only
def test_image_io_var(resx, resy, ext):
pixel = ti.var(dt=ti.f32, shape = (resx, resy))

@ti.kernel
def draw():
    for i, j in pixel:
        pixel[i, j] = ti.random()

draw()

from tempfile import mkstemp

fn = mkstemp(suffix='.' + ext)[1]
ti.imwrite(pixel, fn)

pixel_r = ti.imread(fn).reshape(resx, resy)

pixel = (pixel.to_numpy() * 255 + 0.5).astype(np.uint8)

assert (pixel_r == pixel).all()
```